File: agentcreds-runtime/src/agentcreds_runtime/a2a.py
"""Agent-to-agent (A2A) identity verification - no MCP server.

When agent A hands a task directly to agent B, A attaches a self-contained
identity header (``AgentCreds-A2A/1.<base64url>``) and B verifies it **offline**.
Unlike the interactive MCP path, A2A is one-shot: the *sender* mints the
proof-of-possession challenge with `audience` set to the receiver, so there is no
session and no challenge to store.

`A2AVerifier` is the receiver-side counterpart to `IdentityEnforcer`: it runs the
same shared policy (revocation, multi-issuer anchor resolution, required argument
binding, audit/ADR) on top of the A2A presentation's audience + rooted + PoP
check. `make_a2a_header` is the sender-side helper.

On-behalf-of over A2A: because there is no verified session to establish *who the
human is*, the human principal must travel with the message and be independently
verifiable by the receiver. `A2AVerifier.authorize_obo` validates a principal
token (e.g. an OIDC ID token) via a `principal_resolver` to derive the
`acting_for`, restoring the confused-deputy protection.
"""

from __future__ import annotations

import base64
import hashlib
from typing import Callable, Mapping, Optional

import agentcreds as ac

from .errors import (
    AccessDenied,
    ChainHop,
    Decision,
    CODE_MALFORMED,
    CODE_PRINCIPAL,
    CODE_REPLAY,
)
from .policy import (
    AnchorResolver,
    PolicyConfig,
    PolicyContext,
    adr_signal_for,
    classify_exception,
)
from .replay import InMemoryReplayGuard, ReplayGuard

# A principal resolver: given a principal token (e.g. an OIDC ID token), return
# the verified human DID to use as `acting_for`, or None if it does not validate.
PrincipalResolver = Callable[[str], Optional[str]]

#: Header name carrying the capability presentation (its value is produced by
#: :func:`make_a2a_header` / ``Presentation.to_a2a_header``).
A2A_HEADER_NAME = "AgentCreds-A2A"
#: Header name carrying the on-behalf-of principal token (the human's verifiable
#: identity), produced by :func:`make_a2a_principal_header`.
A2A_PRINCIPAL_HEADER_NAME = "AgentCreds-A2A-Principal"
_A2A_PRINCIPAL_SCHEME = "AgentCreds-A2A-Principal/1."
#: Header name carrying the holder's literal serialized arguments under
#: `agentcreds-octets-v1` - the bytes it signed. Base64url like the principal header,
#: so arbitrary argument text cannot inject header syntax. Absent under every other
#: profile, which reconstructs the binding string by canonicalizing instead.
A2A_BOUND_ARGS_HEADER_NAME = "AgentCreds-A2A-Bound-Args"
#: Header name by which a sender declares the canonicalization profile it computed the
#: binding under. The MCP transport has carried this since -02 §1.3; A2A did not, so a
#: profile disagreement over A2A surfaced as `possession_failed` with no diagnosis.
A2A_CANON_PROFILE_HEADER_NAME = "AgentCreds-A2A-Canon-Profile"


def make_a2a_principal_header(principal_token: str) -> str:
    """Wrap a human principal token (e.g. an OIDC ID token) as the on-behalf-of
    header value: ``AgentCreds-A2A-Principal/1.<base64url(token)>``.

    Sent alongside the capability header (:func:`make_a2a_header`) so the receiver
    can validate *who the human is* independently of what the sender claims.
    """
    body = base64.urlsafe_b64encode(principal_token.encode("utf-8")).rstrip(b"=").decode()
    return f"{_A2A_PRINCIPAL_SCHEME}{body}"


def parse_a2a_principal_header(value: str) -> str:
    """Decode an ``AgentCreds-A2A-Principal/1....`` header back to the principal token.

    # Raises
    ValueError if the scheme prefix is missing or the body is not valid base64url.
    """
    if not value.startswith(_A2A_PRINCIPAL_SCHEME):
        raise ValueError(f"missing '{_A2A_PRINCIPAL_SCHEME}' scheme prefix")
    body = value[len(_A2A_PRINCIPAL_SCHEME):]
    padded = body + "=" * (-len(body) % 4)
    return base64.urlsafe_b64decode(padded.encode()).decode("utf-8")


def make_a2a_bound_args_header(bound_arguments: str) -> str:
    """Wrap the holder's serialized arguments as a base64url header value."""
    return base64.urlsafe_b64encode(bound_arguments.encode("utf-8")).rstrip(b"=").decode()


def parse_a2a_bound_args_header(value: str) -> str:
    """Decode an ``AgentCreds-A2A-Bound-Args`` header back to the serialized arguments.

    # Raises
    ValueError if the body is not valid base64url or not UTF-8.
    """
    padded = value + "=" * (-len(value) % 4)
    return base64.urlsafe_b64decode(padded.encode()).decode("utf-8")


def make_a2a_header(
    token: "ac.DelegationToken",
    credential: "ac.CapabilityCredential",
    leaf_agent: "ac.AgentIdentity",
    *,
    audience: str,
    action: "Optional[ac.Action]" = None,
) -> str:
    """Sender side: mint an A2A identity header bound to `audience` (the receiver).

    Pass `action` (with the same tool/canonical-parameters/resource the receiver
    will see) to bind the proof to this exact request, so a captured header cannot
    be reused for different arguments.
    """
    challenge = ac.PopChallenge(audience)
    if action is not None:
        challenge = challenge.with_request_binding(action.request_binding())
    presentation = ac.Presentation.create(token, credential, challenge, leaf_agent)
    return presentation.to_a2a_header()


def make_a2a_envelope(
    token: "ac.DelegationToken",
    credential: "ac.CapabilityCredential",
    leaf_agent: "ac.AgentIdentity",
    *,
    audience: str,
    action: "Optional[ac.Action]" = None,
    principal_token: "Optional[str]" = None,
    bound_arguments: "Optional[str]" = None,
    canonicalization_profile: "Optional[str]" = None,
) -> "dict[str, str]":
    """Build the A2A wire envelope as a header-name -> value mapping, ready to merge
    into the outgoing request/task headers.

    Always contains ``AgentCreds-A2A`` (the capability). When `principal_token` is
    given (on-behalf-of), it also contains ``AgentCreds-A2A-Principal`` carrying the
    human's verifiable identity. When `bound_arguments` is given (the
    `agentcreds-octets-v1` profile), it also contains ``AgentCreds-A2A-Bound-Args``
    carrying the exact bytes the holder signed. The receiver passes the headers
    straight to :meth:`A2AVerifier.authorize_envelope`.
    """
    envelope = {
        A2A_HEADER_NAME: make_a2a_header(
            token, credential, leaf_agent, audience=audience, action=action
        )
    }
    if canonicalization_profile is not None:
        envelope[A2A_CANON_PROFILE_HEADER_NAME] = canonicalization_profile
    if bound_arguments is not None:
        envelope[A2A_BOUND_ARGS_HEADER_NAME] = make_a2a_bound_args_header(bound_arguments)
    if principal_token is not None:
        envelope[A2A_PRINCIPAL_HEADER_NAME] = make_a2a_principal_header(principal_token)
    return envelope


def principal_resolver_from_oidc(
    provider: "ac.OidcProvider", *, expected_agent_did: Optional[str] = None
) -> PrincipalResolver:
    """Build a :data:`PrincipalResolver` that validates an OIDC ID token (or RFC
    8693 OBO token) against `provider` and returns the human's DID.

    Pass `expected_agent_did` to require the token's `act` actor (if present) to be
    the sending agent. Returns None on any validation failure.
    """

    def resolve(principal_token: str) -> Optional[str]:
        try:
            human = provider.validate_id_token(principal_token, expected_agent_did)
        except ac.AgentCredsError:
            return None
        return human.human_identity().did

    return resolve


class A2AVerifier(PolicyContext):
    """Receiver-side verifier for A2A identity headers.

    Parameters
    ----------
    audience:
        This receiver's identity (e.g. the task URI or the receiving agent's DID).
        An accepted header's embedded audience must equal this, so a header minted
        for another receiver cannot be replayed here. Required.
    anchor / anchor_for:
        The single trust anchor whose credentials this receiver accepts, or an
        `AnchorResolver` for multiple issuers. Exactly one is required.
    config:
        A :class:`~agentcreds_runtime.policy.PolicyConfig` with the shared policy
        knobs (freshness, revocation, policy / usage / approval gates and their
        fail-open posture, argument binding, audit / ADR wiring). Keep
        `max_age_secs` short: within it a header can be replayed unless a
        `replay_guard` is configured. Defaults to the safe `PolicyConfig()`.
    principal_resolver:
        For on-behalf-of: validates a principal token into a human DID (see
        :meth:`authorize_obo` and :func:`principal_resolver_from_oidc`).
    replay_guard / enable_replay_protection:
        Makes each header **single-use**, closing the residual same-request replay
        window inherent to callback-free A2A. **On by default** with an in-process
        `InMemoryReplayGuard`. For more than one receiver replica, pass a shared
        `RedisReplayGuard` (so a replay is caught across replicas). Set
        `enable_replay_protection=False` to turn it off (e.g. if you dedupe
        upstream). A custom `replay_guard` always wins.
    """

    def __init__(
        self,
        *,
        audience: str,
        anchor: "Optional[ac.TrustAnchor]" = None,
        anchor_for: Optional[AnchorResolver] = None,
        config: Optional[PolicyConfig] = None,
        principal_resolver: Optional[PrincipalResolver] = None,
        replay_guard: Optional[ReplayGuard] = None,
        enable_replay_protection: bool = True,
    ):
        super().__init__(anchor, anchor_for=anchor_for, config=config)
        if not audience:
            raise ValueError("A2AVerifier requires an `audience` (this receiver's identity)")
        self._audience = audience
        self._principal_resolver = principal_resolver
        # Single-use protection is on by default (A2A senders mint a fresh
        # challenge per message, so legitimate headers are always distinct).
        if replay_guard is not None:
            self._replay_guard = replay_guard
        elif enable_replay_protection:
            self._replay_guard = InMemoryReplayGuard(ttl_secs=max(self._max_age + 60, 120))
        else:
            self._replay_guard = None

    def authorize(
        self,
        header: str,
        tool: str,
        arguments: object,
        *,
        resource: Optional[str] = None,
        acting_for: Optional[str] = None,
        bound_arguments: Optional[str] = None,
        canonicalization_profile: Optional[str] = None,
    ) -> Decision:
        """Verify an A2A identity header for one task/tool. Returns a `Decision`;
        never raises for an authorization failure.

        `acting_for` (the verified human DID) is supplied by the caller for
        on-behalf-of tokens - use :meth:`authorize_obo` to derive it from a
        principal token instead. `resource` is held to the token's resource scope.

        `bound_arguments` is the holder's literal serialized arguments, required only
        under `agentcreds-octets-v1`; every other profile ignores it.
        """
        args_str = self._initial_args_str(arguments)

        try:
            presentation = ac.Presentation.from_a2a_header(header)
        except Exception as exc:  # malformed header / wrong bytes
            return self._finish(
                self._audience, tool, args_str,
                Decision(False, CODE_MALFORMED, f"could not parse A2A header: {exc}", tool),
                signal="signature_invalid",
            )

        # Same order as the MCP path: a binding computed under a different profile is a
        # profile failure, not a possession failure, and must not be reported as one.
        if (denied := self._deny_if_canon_mismatch(
            presentation, self._audience, tool, args_str, canonicalization_profile,
        )) is not None:
            return denied

        # Under `agentcreds-octets-v1` the binding string becomes the bytes the holder
        # signed and carried, so nothing below re-serializes the arguments.
        denied, args_str, signed_arguments = self._resolve_bound_octets(
            presentation, self._audience, tool, args_str, arguments, bound_arguments,
        )
        if denied is not None:
            return denied

        if (denied := self._deny_if_unbound(presentation, self._audience, tool, args_str)) is not None:
            return denied

        anchor, denied = self._resolve_anchor(presentation, self._audience, tool, args_str)
        if denied is not None:
            return denied

        action = ac.Action(tool, args_str, resource=resource, acting_for=acting_for)
        try:
            # verify_a2a additionally requires the embedded audience == self._audience.
            presentation.verify_a2a(action, anchor, self._audience, self._max_age)
        except ac.AgentCredsError as exc:
            return self._finish(
                self._audience, tool, args_str,
                Decision(False, classify_exception(exc), str(exc), tool),
                token=presentation.token, credential=presentation.credential, signal=adr_signal_for(exc),
            )

        # Single-use enforcement: a verified header may be admitted only once
        # within the freshness window. Recorded only after verify, so invalid
        # headers cannot poison the guard.
        if self._replay_guard is not None:
            key = hashlib.sha256(header.encode("utf-8")).hexdigest()
            if not self._replay_guard.record_if_new(key):
                return self._finish(
                    self._audience, tool, args_str,
                    Decision(False, CODE_REPLAY, "this A2A header has already been used", tool),
                    token=presentation.token, credential=presentation.credential, signal="other",
                )

        if (denied := self._deny_if_revoked(presentation, self._audience, tool, args_str)) is not None:
            return denied

        chain = [
            ChainHop(e.depth, e.agent_did, list(e.tools), e.budget_usd)
            for e in presentation.token.chain().entries
        ]

        # Final contextual gate (same shared hook as the MCP path); `acting_for` is
        # the independently-verified human principal for on-behalf-of messages.
        if (denied := self._deny_if_policy(
            self._audience, tool, args_str,
            principal=acting_for, arguments=arguments, resource=resource,
            chain=chain, token=presentation.token, credential=presentation.credential,
        )) is not None:
            return denied

        # Stateful usage gate (rate/quota + spend).
        if (denied := self._deny_if_usage(
            self._audience, tool, args_str,
            principal=acting_for, arguments=arguments, resource=resource,
            chain=chain, token=presentation.token, credential=presentation.credential,
        )) is not None:
            return denied

        # Human-in-the-loop step-up - runs last; blocks until approved or denied/timeout.
        if (denied := self._hold_for_approval(
            self._audience, tool, args_str,
            principal=acting_for, arguments=arguments, resource=resource,
            chain=chain, token=presentation.token, credential=presentation.credential, anchor=anchor,
        )) is not None:
            return denied

        return self._finish(
            self._audience, tool, args_str,
            Decision(True, None, None, tool, chain, bound_arguments=signed_arguments),
            token=presentation.token, credential=presentation.credential,
        )

    def authorize_obo(
        self,
        header: str,
        principal_token: str,
        tool: str,
        arguments: object,
        *,
        resource: Optional[str] = None,
        bound_arguments: Optional[str] = None,
        canonicalization_profile: Optional[str] = None,
    ) -> Decision:
        """Verify an on-behalf-of A2A message: independently validate the human's
        `principal_token` (via the configured `principal_resolver`) to derive the
        `acting_for`, then run :meth:`authorize`. The core then enforces that the
        capability token's bound principal matches the verified human - restoring
        the confused-deputy protection that the MCP session binding provides.
        """
        if self._principal_resolver is None:
            raise ValueError("authorize_obo requires a `principal_resolver` on the verifier")
        acting_for = self._principal_resolver(principal_token)
        if acting_for is None:
            return self._finish(
                self._audience, tool, self._initial_args_str(arguments),
                Decision(False, CODE_PRINCIPAL, "could not verify the on-behalf-of principal", tool),
                signal="other",
            )
        return self.authorize(header, tool, arguments, resource=resource,
                              acting_for=acting_for, bound_arguments=bound_arguments,
                              canonicalization_profile=canonicalization_profile)

    def authorize_envelope(
        self,
        headers: "Mapping[str, str]",
        tool: str,
        arguments: object,
        *,
        resource: Optional[str] = None,
    ) -> Decision:
        """Verify a request from its A2A wire envelope (a header-name -> value
        mapping, e.g. the incoming HTTP/task headers built by
        :func:`make_a2a_envelope`).

        Reads the ``AgentCreds-A2A`` capability header and, if present, the
        ``AgentCreds-A2A-Principal`` on-behalf-of header - dispatching to
        :meth:`authorize_obo` when a principal is carried, or :meth:`authorize`
        otherwise. Header-name lookup is case-insensitive.
        """
        lookup = {k.lower(): v for k, v in headers.items()}
        capability = lookup.get(A2A_HEADER_NAME.lower())
        if capability is None:
            return self._finish(
                self._audience, tool, self._initial_args_str(arguments),
                Decision(False, CODE_MALFORMED,
                         f"envelope missing the '{A2A_HEADER_NAME}' header", tool),
                signal="signature_invalid",
            )

        bound_arguments = None
        bound_header = lookup.get(A2A_BOUND_ARGS_HEADER_NAME.lower())
        if bound_header is not None:
            try:
                bound_arguments = parse_a2a_bound_args_header(bound_header)
            except (ValueError, UnicodeDecodeError) as exc:
                return self._finish(
                    self._audience, tool, self._initial_args_str(arguments),
                    Decision(False, CODE_MALFORMED,
                             f"bad {A2A_BOUND_ARGS_HEADER_NAME} header: {exc}", tool),
                    signal="signature_invalid",
                )

        declared_profile = lookup.get(A2A_CANON_PROFILE_HEADER_NAME.lower())

        principal_header = lookup.get(A2A_PRINCIPAL_HEADER_NAME.lower())
        if principal_header is None:
            return self.authorize(capability, tool, arguments, resource=resource,
                                  bound_arguments=bound_arguments,
                                  canonicalization_profile=declared_profile)
        try:
            principal_token = parse_a2a_principal_header(principal_header)
        except ValueError as exc:
            return self._finish(
                self._audience, tool, self._initial_args_str(arguments),
                Decision(False, CODE_MALFORMED, f"bad principal header: {exc}", tool),
                signal="signature_invalid",
            )
        return self.authorize_obo(capability, principal_token, tool, arguments,
                                  resource=resource, bound_arguments=bound_arguments,
                                  canonicalization_profile=declared_profile)

    def enforce(
        self,
        header: str,
        tool: str,
        arguments: object,
        *,
        resource: Optional[str] = None,
        acting_for: Optional[str] = None,
    ):
        """Like `authorize`, but raise `AccessDenied` on a deny and return the
        verified chain on allow."""
        decision = self.authorize(header, tool, arguments, resource=resource, acting_for=acting_for)
        if decision.denied:
            raise AccessDenied(decision)
        return decision.chain

File: agentcreds-core/src/a2a.rs
//! A2A (agent-to-agent) identity header.
//!
//! Attaches a signed capability [`Presentation`] to an A2A task delegation so
//! the receiving agent can verify the caller's authority **offline** - no
//! callback to the issuer. The header is a compact, self-contained string
//! suitable for an HTTP header value or an A2A v1.0 Agent Card extension field.
//!
//! Wire form: `AgentCreds-A2A/1.<base64url(CBOR(Presentation))>`.
//!
//! ## Callback-free verification
//!
//! A normal [`Presentation`] is checked against a challenge the verifier issued
//! interactively. A2A task delegation is one-shot, so the **sender** mints the
//! challenge with `audience` set to the receiver and signs proof-of-possession
//! over it. The receiver verifies the embedded proof, requires the audience to
//! be itself, and enforces a freshness window. Within that window and audience
//! a captured header can be replayed, so keep `max_age_secs` short and the
//! audience specific (e.g. the task URI or the receiving agent's DID).
//!
//! ```rust
//! use agentcreds_core::prelude::*;
//!
//! // Caller side: mint a presentation bound to the receiver as audience.
//! let anchor = TrustAnchor::generate()?;
//! let agent  = AgentIdentity::create(DidMethod::Key, None)?;
//! let claims = CapabilityClaims::new(vec!["tool:search".into()], 1, 3600);
//! let vc = CapabilityCredential::issue(&anchor, agent.did(), claims, None)?;
//! let scope = Scope::with_budget_and_depth(vec!["tool:search".into()], None, 1);
//! let token = DelegationToken::mint(&vc, scope, 300, &agent)?;
//!
//! let receiver = "a2a://orders.example/agent";
//! let challenge = PopChallenge::new(Some(receiver.into()));
//! let presentation = Presentation::create(token, vc, &challenge, &agent)?;
//! let header = presentation.to_a2a_header()?;
//!
//! // Receiver side: verify offline, requiring the audience to be itself.
//! let received = Presentation::from_a2a_header(&header)?;
//! received.verify_a2a(&Action::new("tool:search", "q=ok"), &anchor, receiver, 60)?;
//! # Ok::<(), agentcreds_core::error::AgentCredsError>(())
//! ```

use base64ct::{Base64Url, Encoding};

use crate::delegation::Action;
use crate::did::TrustAnchor;
use crate::pop::Presentation;
use crate::{error::AgentCredsError, Result};

/// Scheme/version prefix for the A2A header wire form.
const A2A_SCHEME: &str = "AgentCreds-A2A/1.";

impl Presentation {
    /// Encode this presentation as an A2A identity header string.
    ///
    /// # Errors
    /// Propagates a CBOR serialization error.
    pub fn to_a2a_header(&self) -> Result<String> {
        let cbor = self.to_cbor()?;
        Ok(format!("{A2A_SCHEME}{}", Base64Url::encode_string(&cbor)))
    }

    /// Decode an A2A identity header string back into a presentation.
    ///
    /// # Errors
    /// `OutOfBounds` if the scheme prefix is missing or the body is not valid
    /// base64url; a deserialization error if the CBOR is malformed.
    pub fn from_a2a_header(header: &str) -> Result<Self> {
        let body = header
            .strip_prefix(A2A_SCHEME)
            .ok_or_else(|| AgentCredsError::OutOfBounds {
                field: "a2a_header",
                detail: format!("missing '{A2A_SCHEME}' scheme prefix"),
            })?;
        let cbor = Base64Url::decode_vec(body).map_err(|e| AgentCredsError::OutOfBounds {
            field: "a2a_header",
            detail: format!("invalid base64url body: {e}"),
        })?;
        Presentation::from_cbor(&cbor)
    }

    /// Verify an A2A presentation as the receiving agent.
    ///
    /// In addition to the full [`Presentation::verify`] checks (anchor-rooted
    /// token plus proof of possession), this requires the embedded challenge's
    /// audience to equal `expected_audience` (this verifier), so a header minted
    /// for one receiver cannot be replayed against another.
    ///
    /// # Errors
    /// `ProofOfPossessionFailed` if the audience does not match, plus any error
    /// from the underlying presentation verification.
    pub fn verify_a2a(
        &self,
        action: &Action,
        anchor: &TrustAnchor,
        expected_audience: &str,
        max_age_secs: i64,
    ) -> Result<()> {
        match self.proof.challenge.audience.as_deref() {
            Some(a) if a == expected_audience => {}
            _ => {
                return Err(AgentCredsError::ProofOfPossessionFailed {
                    reason: "a2a header audience does not match this verifier".into(),
                })
            }
        }
        let challenge = self.proof.challenge.clone();
        self.verify(action, anchor, &challenge, max_age_secs)
    }
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
mod tests {
    use crate::delegation::{Action, DelegationToken, Scope};
    use crate::did::{AgentIdentity, DidMethod, TrustAnchor};
    use crate::pop::{PopChallenge, Presentation};
    use crate::vc::{CapabilityClaims, CapabilityCredential};

    const RECEIVER: &str = "a2a://orders.example/agent";

    fn presentation_for(audience: Option<&str>) -> (TrustAnchor, Presentation) {
        presentation_aged(audience, chrono::Duration::zero())
    }

    /// Like [`presentation_for`], but the PoP challenge is back-dated by `age_back`
    /// (set *before* signing, so the digest stays valid). The success-path test
    /// uses this with a wide freshness window so the assertion turns on the
    /// round-trip + signature/binding checks, **not** on the wall clock holding
    /// still: a backward time step between issue and verify (e.g. an NTP step under
    /// load) must not be able to trip the future-skew bound. The freshness bounds
    /// themselves are covered by the dedicated tests in `pop`.
    fn presentation_aged(
        audience: Option<&str>,
        age_back: chrono::Duration,
    ) -> (TrustAnchor, Presentation) {
        let anchor = TrustAnchor::generate().unwrap();
        let agent = AgentIdentity::create(DidMethod::Key, None).unwrap();
        let claims = CapabilityClaims::new(vec!["tool:search".into()], 1, 3600);
        let vc = CapabilityCredential::issue(&anchor, agent.did(), claims, None).unwrap();
        let scope = Scope::with_budget_and_depth(vec!["tool:search".into()], None, 1);
        let token = DelegationToken::mint(&vc, scope, 300, &agent).unwrap();
        let mut challenge = PopChallenge::new(audience.map(str::to_string));
        challenge.issued_at = challenge.issued_at - age_back;
        let presentation = Presentation::create(token, vc, &challenge, &agent).unwrap();
        (anchor, presentation)
    }

    #[test]
    fn header_round_trip_and_verify() {
        // Back-date the challenge 60s and verify within a 1-hour window: the result
        // turns on the round-trip + crypto, and tolerates ~65s of backward clock
        // movement between issue and verify so a wall-clock step can't flake it.
        let (anchor, presentation) =
            presentation_aged(Some(RECEIVER), chrono::Duration::seconds(60));
        let header = presentation.to_a2a_header().unwrap();
        assert!(header.starts_with("AgentCreds-A2A/1."));

        let received = Presentation::from_a2a_header(&header).unwrap();
        assert!(received
            .verify_a2a(&Action::new("tool:search", "q=ok"), &anchor, RECEIVER, 3600)
            .is_ok());
    }

    #[test]
    fn wrong_audience_rejected() {
        let (anchor, presentation) = presentation_for(Some(RECEIVER));
        let header = presentation.to_a2a_header().unwrap();
        let received = Presentation::from_a2a_header(&header).unwrap();
        // A different receiver must not accept a header minted for RECEIVER.
        assert!(received
            .verify_a2a(
                &Action::new("tool:search", "q=ok"),
                &anchor,
                "a2a://evil.example/agent",
                60,
            )
            .is_err());
    }

    #[test]
    fn missing_audience_rejected() {
        let (anchor, presentation) = presentation_for(None);
        let header = presentation.to_a2a_header().unwrap();
        let received = Presentation::from_a2a_header(&header).unwrap();
        assert!(received
            .verify_a2a(&Action::new("tool:search", "q=ok"), &anchor, RECEIVER, 60)
            .is_err());
    }

    #[test]
    fn missing_scheme_prefix_rejected() {
        assert!(Presentation::from_a2a_header("not-a-header").is_err());
    }

    #[test]
    fn tampered_header_body_rejected() {
        let (_anchor, presentation) = presentation_for(Some(RECEIVER));
        let header = presentation.to_a2a_header().unwrap();
        // Corrupt a character in the base64url body.
        let mut chars: Vec<char> = header.chars().collect();
        let last = chars.len() - 1;
        chars[last] = if chars[last] == 'A' { 'B' } else { 'A' };
        let tampered: String = chars.into_iter().collect();
        // Either decoding or verification must fail - never silently accept.
        let parsed = Presentation::from_a2a_header(&tampered);
        assert!(
            parsed.is_err() || {
                let anchor = TrustAnchor::generate().unwrap();
                parsed
                    .unwrap()
                    .verify_a2a(&Action::new("tool:search", "q=ok"), &anchor, RECEIVER, 60)
                    .is_err()
            }
        );
    }
}
